mgr/dashboard: add smb service management support

Pegonzal opened the pull request:
![image](https://github.com/ceph/ceph/assets/106810838/3541a240-3636-4910-ab81-fe3f2a1d30b6)


[screen-capture (48).webm](https://github.com/ceph/ceph/assets/106810838/3d2a9f96-7bc3-4951-a898-a29010beed55)

Fixes: https://tracker.ceph.com/issues/65681





<!--
  - Please give your pull request a title like

      [component]: [short description]

  - Please use this format for each git commit message:

      [component]: [short description]

      [A longer multiline description]

      Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
      Signed-off-by: [Your Name] <[your email]>

    For examples, use "git log".
-->

## Contribution Guidelines
- To sign and title your commits, please refer to [Submitting Patches to Ceph](https://github.com/ceph/ceph/blob/main/SubmittingPatches.rst).

- If you are submitting a fix for a stable branch (e.g. "quincy"), please refer to [Submitting Patches to Ceph - Backports](https://github.com/ceph/ceph/blob/master/SubmittingPatches-backports.rst) for the proper workflow.

- When filling out the below checklist, you may click boxes directly in the GitHub web UI.  When entering or editing the entire PR message in the GitHub web UI editor, you may also select a checklist item by adding an `x` between the brackets: `[x]`.  Spaces and capitalization matter when checking off items this way.

## Checklist
- Tracker (select at least one)
  - [x] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [ ] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [x] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [ ] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
- `jenkins test rook e2e`
</details>

avanthakkar reviewed (commented):
LGTM functionally!
Can you please also add a unit test for smb service creation

cloudbehl commented:
@Pegonzal Thanks for PR.

1. Can we bring required items on top.
2. Can you add example placeholders for all items.
3. `Config uri` => `Config URI`
4. ceph users could be selectable rather than entering manually. 
5. Features we have only one(`domain`), can that be a list and can `domain` feature be enabled by default? @phlogistonjohn can you help suggest, if its a ip to have otherwise we can remove the field? 


@phlogistonjohn can you help, what things from your perspective could be made easier for user by automating or having preselected values?

phlogistonjohn commented:
Hi @cloudbehl this feature is really meant to support the (recently merged) smb mgr module and is not really meant to end users to interact with directly. @Pegonzal needs to make the existing service spec support complete but should only be done minimally IMO. I wouldn't go too far in making it "friendly" just do enough to be complete. If you are an "expert" in the samba-containers project you could use this service spec directly but that's a vanishingly small number of people.

If you want, features can be a multi-select or group of checkboxes (or whatever your framework calls it). The enumeration for 'features' is currently either `domain` or nothing but future versions will likely add `metrics` and `cluster` (names not set in stone). But it'll always be a list of "enums". The default should be the equivalent of an "empty list". The features list tells cephadm what specific grouping of containers will be needed.

The smb mgr module PR also had to add an additional field. `user_sources` ([link](https://github.com/ceph/ceph/blob/04416f48efe02891b265c07b4f38b46e422bbef9/src/python-common/ceph/deployment/service_spec.py#L2578)) that works similarly to `join_sources` - it takes a list of URI values that sambacc will use to download user configuration from.

Pegonzal commented:
> Can you please also add a unit test for smb service creation

Added it

> Can we bring required items on top.
> Can you add example placeholders for all items.
> Config uri => Config URI

All addressed

> ceph users could be selectable rather than entering manually.

Created a tracker to add this new feature in the forms that might need it https://tracker.ceph.com/issues/66032

Also added the `user_sources` field and changed the `features` field to use checkboxes, so the user can select any if needed.

Pegonzal commented:
jenkins test dashboard

Pegonzal commented:
jenkins test dashboard

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -395,6 +395,161 @@
+                     placeholder="rados://.smb/foo/scc.toml"
+                     i18n-placeholder>
+              <span class="invalid-feedback"
+                    *ngIf="serviceForm.showError('config_uri', frm, 'required')"
```
good to have validation for url based on `Supported URI schemes include http:, https:, rados:, and rados:mon-config-key:.`

cloudbehl commented on src/pybind/mgr/dashboard/frontend/src/app/ceph/cluster/services/service-form/service-form.component.html:
```diff
@@ -395,6 +395,161 @@
+            <label class="cd-col-form-label required"
+                   for="cluster_id"
+                   i18n>
+              Cluster id
```
```suggestion
              Cluster name
```

cloudbehl replied:
Lets call it name, ID is confusing.

cloudbehl replied:
or something like this: Cluster ID(Name)

Pegonzal replied:
Hey @cloudbehl , I get your point, but there is already a helper to give additional info for the field in case that something might not be clear enough, even when we know this feature will be mostly useful for advanced users on the topic.
Currently the naming of every field follows both the backend implementation as well as the documentation names:
https://docs.ceph.com/en/latest/cephadm/services/smb/#service-spec-options

In my opinion chaning the name of things should be done carefully and not arbitrarely as this is the easiest way to make things confusing and unmantainable

cloudbehl reviewed (approved):
minor comments, mostly looks good.

nizamial09 reviewed (approved):
LGTM! can you add an e2e test in the cephadm dashboard e2e suite for the service creation? Since we don't expect users to use this interface to frequently there's a good chance that we can miss out if this is broken unintentionally.

Pegonzal commented:
jenkins test make check

Pegonzal commented:
jenkins test dashboard

Pegonzal commented:
jenkins test dashboard cephadm

Pegonzal commented:
jenkins test dashboard

Pegonzal commented:
jenkins test dashboard
